File: Backtracking/NQueensII.py
# LeetCode 52 - N-Queens II
# Time Complexity: O(n!) | Space Complexity: O(n)


class Solution:
    def totalNQueens(self, n: int) -> int:
        columns = [False] * n
        diagonals = [False] * (n * 2)
        anti_diagonals = [False] * (n * 2)
        answer = 0

        def backtrack(row: int) -> None:
            nonlocal answer
            if row == n:
                answer += 1
                return

            for col in range(n):
                diagonal_index = row + col
                anti_diagonal_index = row - col + n
                if columns[col] or diagonals[diagonal_index] or anti_diagonals[anti_diagonal_index]:
                    continue

                columns[col] = True
                diagonals[diagonal_index] = True
                anti_diagonals[anti_diagonal_index] = True
                backtrack(row + 1)
                columns[col] = False
                diagonals[diagonal_index] = False
                anti_diagonals[anti_diagonal_index] = False

        backtrack(0)
        return answer

File: Backtracking/NQueensII.java
// LeetCode 52 - N-Queens II
// Time Complexity: O(n!) | Space Complexity: O(n)
public class NQueensII {
    private int size;
    private int answer;
    private final boolean[] columns = new boolean[10];
    private final boolean[] diagonals = new boolean[20];
    private final boolean[] antiDiagonals = new boolean[20];

    public int totalNQueens(int n) {
        size = n;
        answer = 0;
        backtrack(0);
        return answer;
    }

    private void backtrack(int row) {
        if (row == size) {
            answer++;
            return;
        }

        for (int col = 0; col < size; col++) {
            int diagonalIndex = row + col;
            int antiDiagonalIndex = row - col + size;
            if (columns[col] || diagonals[diagonalIndex] || antiDiagonals[antiDiagonalIndex]) {
                continue;
            }

            columns[col] = true;
            diagonals[diagonalIndex] = true;
            antiDiagonals[antiDiagonalIndex] = true;
            backtrack(row + 1);
            columns[col] = false;
            diagonals[diagonalIndex] = false;
            antiDiagonals[antiDiagonalIndex] = false;
        }
    }
}
